teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+
+Pros:
+* "fixes" PipelineResources
+* Uses concepts we already have in Tekton but upgrades them
+
+Cons:
+* Not clear what the idea of a PipelineResource is really giving us if it's just a wrapper for Tasks
+* If you want to use 2 Tasks together, you'll have to make a PipelineResource type for at least one of them
+* Only helps us with some scheduling problems (e.g. doesn't help with parallel tasks or finally task execution)
+
+Related:
+* [Specializing Tasks: Visions and Goals](https://docs.google.com/document/d/1G2QbpiMUHSs4LOqcNaIRswcdvoy8n7XuhTV8tXdcE7A/edit)
+* [Specializing Tasks: Possible Designs](https://docs.google.com/document/d/1p8zq_wkAcwr1l5BpNQDyNjgWngOtnEhCYEpcNKMHvG4/edit)
+
+### Automagically combine Tasks based on workspace use
+
+In this option we could leave Pipelines as they are, but at runtime instead of mapping a Task to a pod, we could decide
+what belongs in what pod based on workspace usage.
+
+In the example below, `get-source`, `run-unit-tests` and `upload-results` are all at least one of the two workspaces
+so they will be executed as one pod, while `update-slack` would be run as a separate pod:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+ params:
+  - name: url
+    value: https://github.com/tektoncd/pipeline.git
+  - name: revision
+    value: v0.11.3
+ workspaces:
+  - name: source-code
+  - name: test-results
+ tasks:
+ - name: get-source
+   workspaces:
+   - name: source-code
+     workspace: source-code
+   taskRef:
+     name: git-clone
+   params:
+   - name: url
+      value: $(params.url)
+    - name: revision
+      value: $(params.revision)
+ - name: run-unit-tests
+   runAfter: get-source
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   taskRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+finally:
+- name: update-slack
+  params:
+  - name: message
+    value: "Tests completed with $(tasks.run-unit-tests.status) status"
+```
+
+Possible tweak: we could do this scheduling only when
+[a Task requires a workspace `from` another Task](https://github.com/tektoncd/pipeline/issues/3109).
+
+Pros:
+* Doesn't require any changes for Pipeline or Task authors
+
+Cons:
+* Will need to update our entrypoint logic to allow for steps running in parallel
+* Doesn't give as much flexibility as being explicit
+  * This functionality might not even be desirable for folks who want to make use of multiple nodes
+
+### Introduce scheduling rules to pipeline
+
+In these options, we add some syntax that allows Pipeline authors to express how they want Tasks to be executed.
+
+#### Add "grouping" to tasks in a pipeline
```

Comment:
a related alternative would be to define the "group" as a subpipeline that would be executed as one pod, instead of creating a new way to group tasks

Reply:
thats a great idea! i wonder how we'd handle it if that pipeline itself had a sub pipeline 🤔  i guess we could either not allow it, or just put it all on the same pod in that case?

Reply:
for reusability, we could allow nesting and we execute all in the same pod

Reply:
@jerop id definitely like it to be - tho i do wonder how custom tasks fit in if you run an entire pipeline in a pod (e.g. are you still creating a CRD object and waiting for it... but from the pod?) I'm guessing that if we do eventually support the entire pipeline being run in a pod, this might be one of the last things we support